modules/apps/callbacks/callbacks_test.go

```diff
@@ -0,0 +1,262 @@
+
+	switch callbackType {
+	case "none":
+		s.Require().Len(sourceCounters, 0)
+		s.Require().Len(destCounters, 0)
+
+	case types.CallbackTriggerSendPacket:
+		s.Require().Len(sourceCounters, 1)
+		s.Require().Equal(1, sourceCounters[types.CallbackTriggerSendPacket])
+
+	case types.CallbackTriggerAcknowledgementPacket:
+		s.Require().Len(sourceCounters, 2)
+		s.Require().Equal(1, sourceCounters[types.CallbackTriggerSendPacket])
+		s.Require().Equal(1, sourceCounters[types.CallbackTriggerAcknowledgementPacket])
+
+		s.Require().Len(destCounters, 0)
+
+	case types.CallbackTriggerReceivePacket:
+		s.Require().Len(sourceCounters, 0)
+		s.Require().Len(destCounters, 1)
+		s.Require().Equal(1, destCounters[types.CallbackTriggerReceivePacket])
+
+	case types.CallbackTriggerTimeoutPacket:
+		s.Require().Len(sourceCounters, 2)
+		s.Require().Equal(1, sourceCounters[types.CallbackTriggerSendPacket])
+		s.Require().Equal(1, sourceCounters[types.CallbackTriggerTimeoutPacket])
+
+		s.Require().Len(destCounters, 0)
+
+	default:
+		s.FailNow(fmt.Sprintf("invalid callback type %s", callbackType))
+	}
+}
+
+func TestIBCCallbacksTestSuite(t *testing.T) {
+	suite.Run(t, new(CallbacksTestSuite))
+}
+
+// AssertHasExecutedExpectedCallbackWithFee checks if only the expected type of callback has been executed
+// and that the expected ics-29 fee has been paid.
+func (s *CallbacksTestSuite) AssertHasExecutedExpectedCallbackWithFee(
+	callbackType types.CallbackTrigger, isSuccessful bool, isTimeout bool,
+	originalSenderBalance sdk.Coins, fee feetypes.Fee,
+) {
+	// Recall that:
+	// - the source chain is chainA
+	// - forward relayer is chainB.SenderAccount
+	// - reverse relayer is chainA.SenderAccount
+	// - The counterparty payee of the forward relayer in chainA is chainB.SenderAccount (as a chainA account)
+
+	if !isTimeout {
+		// check forward relay balance
+		s.Require().Equal(
+			fee.RecvFee,
+			sdk.NewCoins(s.chainA.GetSimApp().BankKeeper.GetBalance(s.chainA.GetContext(), s.chainB.SenderAccount.GetAddress(), ibctesting.TestCoin.Denom)),
+		)
+
+		s.Require().Equal(
+			fee.AckFee.Add(fee.TimeoutFee...), // ack fee paid, timeout fee refunded
+			sdk.NewCoins(
+				s.chainA.GetSimApp().BankKeeper.GetBalance(
+					s.chainA.GetContext(), s.chainA.SenderAccount.GetAddress(),
+					ibctesting.TestCoin.Denom),
+			).Sub(originalSenderBalance[0]),
+		)
+	} else {
+		// forward relay balance should be 0
+		s.Require().Equal(
+			sdk.NewCoin(ibctesting.TestCoin.Denom, sdkmath.ZeroInt()),
+			s.chainA.GetSimApp().BankKeeper.GetBalance(s.chainA.GetContext(), s.chainB.SenderAccount.GetAddress(), ibctesting.TestCoin.Denom),
+		)
+
+		// all fees should be returned as sender is the reverse relayer
+		s.Require().Equal(
+			fee.Total(),
+			sdk.NewCoins(
+				s.chainA.GetSimApp().BankKeeper.GetBalance(
+					s.chainA.GetContext(), s.chainA.SenderAccount.GetAddress(),
+					ibctesting.TestCoin.Denom),
+			).Sub(originalSenderBalance[0]),
+		)
+	}
```

Comment:
nit: use `balance` (or some named var) for `sdk.NewCoins()` values and keep `s.Require().Equal(exp, actual)` on one line for readability